bpo-45949: Pure Python freeze module for cross builds (GH-29899)

tiran opened the pull request:
Use `_bootstrap_python` interpreter and pure Python implementation of
`freeze_module` to generate frozen byte code files. Only importlib
bootstrap files are generated with `Programs/_freeze_module`.

This simplifies cross building, as the build system no longer needs a
`_freeze_module` binary. A standard Python installation with same
version is sufficient.

Signed-off-by: Christian Heimes <christian@python.org>

<!-- issue-number: [bpo-45949](https://bugs.python.org/issue45949) -->
https://bugs.python.org/issue45949
<!-- /issue-number -->

tiran committed:
bpo-45949: Pure Python freeze module for cross builds

Use `_bootstrap_python` interpreter and pure Python implementation of
`freeze_module` to generate frozen byte code files. Only importlib
bootstrap files are generated with `Programs/_freeze_module`.

This simplifies cross building, as the build system no longer needs a
`_freeze_module` binary. A standard Python installation with same
version is sufficient.

Signed-off-by: Christian Heimes <christian@python.org>

tiran committed:
Move script next to _freeze_module.c

ericsnowcurrently commented on Makefile.pre.in:
```diff
@@ -1032,6 +1035,33 @@ Python/deepfreeze/frozen_only.c: Python/frozen_modules/frozen_only.h $(DEEPFREEZ
+# 2) deepfreeze modules with external build Python.
+#
+# Cross compiling also creates Programs/_freeze_module and _bootstrap_python
+# artifacts, but does not use or ship them.
```
Thanks for the great comment!

tiran replied:
You are welcome!

ericsnowcurrently commented on Makefile.pre.in:
```diff
@@ -1032,6 +1035,33 @@ Python/deepfreeze/frozen_only.c: Python/frozen_modules/frozen_only.h $(DEEPFREEZ
+#    script is used to test the cross compile code path.
+# 6) deepfreeze modules with _bootstrap_python
+#
+# Cross compile proces:
```
```suggestion
# Cross compile process:
```

tiran replied:
Thanks

ericsnowcurrently commented on Makefile.pre.in:
```diff
@@ -1083,77 +1113,77 @@ Programs/_freeze_module.o: Programs/_freeze_module.c Makefile
-Python/frozen_modules/importlib._bootstrap.h: $(FREEZE_MODULE) Lib/importlib/_bootstrap.py
-	$(FREEZE_MODULE) importlib._bootstrap $(srcdir)/Lib/importlib/_bootstrap.py Python/frozen_modules/importlib._bootstrap.h
+Python/frozen_modules/importlib._bootstrap.h: Lib/importlib/_bootstrap.py $(FREEZE_MODULE_C)
+	$(FREEZE_MODULE_BOOTSTRAP) importlib._bootstrap $(srcdir)/Lib/importlib/_bootstrap.py Python/frozen_modules/importlib._bootstrap.h
```
Why is the dependency always `$(FREEZE_MODULE_C)`?  Shouldn't that be `$(srcdir)/Programs/_freeze_module.py` if `$(FREEZE_MODULE_BOOTSTRAP)` is `$(FREEZE_MODULE_PY)`?

ericsnowcurrently commented on Makefile.pre.in:
```diff
@@ -1083,77 +1113,77 @@ Programs/_freeze_module.o: Programs/_freeze_module.c Makefile
 
+# We manually freeze getpath.py rather than through freeze_modules
+Python/frozen_modules/getpath.h: Modules/getpath.py $(FREEZE_MODULE_C)
+	$(FREEZE_MODULE_BOOTSTRAP) getpath $(srcdir)/Modules/getpath.py Python/frozen_modules/getpath.h
```
Instead of "manually", can this be generated now?  (added to Tools/scripts/freeze_modules.py)

tiran replied:
getpath is not like the others. It would need special handling in ``Tools/scripts/freeze_modules.py``, too. Let's tackle the getpath target in another PR. I only moved existing block to a different location so it is closer to other early bootstrap blocks.

ericsnowcurrently commented on Tools/scripts/freeze_modules.py:
```diff
@@ -580,10 +584,17 @@ def regen_makefile(modules):
-        freeze = (f'$(FREEZE_MODULE) {src.frozenid} '
-                  f'$(srcdir)/{pyfile} {frozen_header}')
+        if src.isbootstrap:
+            freezecmd = '$(FREEZE_MODULE_BOOTSTRAP)'
```
This implies that `make regen-frozen` must always be run for cross-compiling builds.  Also the committed copy of Makefile.pre.in must only be generated for a non-cross-compiling build, since that is the default.

tiran replied:
It does not matter which build flavor is used to regenerate files. All build / cross-build options are autoconf and automake templates.

However I see that the variables are not self-explanatory. I changed the variables and configure script to set four different variables for freezing.

ericsnowcurrently replied:
Thanks for clarifying and for the change.  That helps.

ericsnowcurrently commented on Programs/_freeze_module.py:
```diff
@@ -0,0 +1,69 @@
+reference count > 1. Marshal adds the `FLAG_REF` flag and creates a
+reference `hashtable`.
+"""
+
```
Would you mind adding a short note here (or even a TODO) about the relationship with Tools/freeze/freeze.py?

tiran replied:
Good idea!

ericsnowcurrently commented on Programs/_freeze_module.py:
```diff
@@ -0,0 +1,69 @@
+"""Python implementation of Programs/_freeze_module.c
```
It would also be worth having a note in Programs/_freeze_module.c pointing back to this file.

ericsnowcurrently commented on Programs/_freeze_module.py:
```diff
@@ -0,0 +1,69 @@
+implementation.
+
+The generated byte code is slightly different because
+compile() sets the PyCF_SOURCE_IS_UTF8 flag and objects have a
+reference count > 1. Marshal adds the `FLAG_REF` flag and creates a
+reference `hashtable`.
```
Yeah, this caused me a big headache at one point because on debug builds some of the frozen strings had multiple refs where they only had 1 on non-debug builds, so the resulting marshaled data was different.  We solved that but it was a pain.  Thanks for reminding me. 😄

tiran replied:
It was a painful experience. At first I did not have ``dont_inherit=True`` and the resulting byte code caused bugs. It took me a while to realize that the problem was caused by the missing flags and not the hashtable.

ericsnowcurrently commented on Programs/_freeze_module.py:
```diff
@@ -0,0 +1,69 @@
+    filename = f"<frozen {name}>"
+    # exec == Py_file_input
+    code = compile(text, filename, "exec", optimize=0, dont_inherit=True)
+    # defaults to Py_MARSHAL_VERSION)
```
What does this mean?

ericsnowcurrently reviewed (approved):
Mostly LGTM.

I've left a few minor comments.  The only matter of consequence is about make rule dependencies in the cross-compiling case, as well as running `make regen-frozen` as one of the steps for a cross-compiled build.

I'm approving the PR under the assumption that you'll resolve my concerns before merging. 🙂

tiran committed:
Address review comments

- Fix typo
- Remove confusing comment
- Improve dependency handling for freezing

ericsnowcurrently reviewed (approved):
LGTM
